packages/gatsby/src/redux/reducers/set-request-headers.ts

```diff
@@ -0,0 +1,48 @@
+import reporter from "gatsby-cli/lib/reporter"
+import url from "url"
+
+import { IGatsbyState, ISetDomainRequestHeaders } from "../types"
+
+export const setRequestHeadersReducer = (
+  state: IGatsbyState["requestHeaders"] = new Map(),
+  action: ISetDomainRequestHeaders
+): IGatsbyState["requestHeaders"] => {
+  switch (action.type) {
+    case `SET_REQUEST_HEADERS`: {
+      const { headers, domain, pluginName } = action.payload
+
+      const noHeaders = typeof headers !== `object`
+      const noDomain = typeof domain !== `string`
+
+      if (noHeaders) {
+        reporter.warn(
+          `Plugin ${pluginName} called actions.setRequestHeaders with a headers property that isn't an object.`
+        )
+      }
+
+      if (noDomain) {
+        reporter.warn(
+          `Plugin ${pluginName} called actions.setRequestHeaders with a domain property that isn't a string.`
+        )
+      }
```

Comment:
you need to do this logic inside the public action, reducers shouldn't think about validation